### PR TITLE
osd: avoid for the two copy to cancel each other

### PR DESCRIPTION
**Pull request by myoungwon**
For cache tier, if some head object has two snaps, the two snaps share the same clone object,  
and the clone object was flush/evicted from cache pool,  when a rollback requests and a read 
snap request to these two snaps at the same time will generate two promote requests to the 
same clone object, these two promote requests will generate two copy ops with same src, than
the second copy op will cancel the first copy op by calling cancel_copy and kick_object_context_blocked, 
but after calling kick_object_context_blocked, a new promote request corresponding to first 
copy op will be restarted and generate a new copy op, the new copy op will cancel the second
copy op again, so two promote requests will cancel their copy op each other and run into dead
loop.

Fixes: https://tracker.ceph.com/issues/49409
Fixes: https://tracker.ceph.com/issues/49726

Signed-off-by: YuanXin <yuanxin@didiglobal.com>
Signed-off-by: mychoxin <mychoxin@gmail.com>
Signed-off-by: Myoungwon Oh <myoungwon.oh@samsung.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by myoungwon** [comment]
> I think I found the root cause. Can you double-check this?
This seems like the sequence inversion because kick_object_context_blocked is moved to cancel_copy_ops().

> Please look at the following. During copy-from op in ceph_test_rados, it also issues the stat op to check racing read. So,

>copy-from to OID A
stat to OID A
The osd receives copy-from, then issue copy request to the target
Stat is received, but it is blocked.
Something happens, so recovery is triggered
cancel_copy_ops() is called, so copy-from is canceled, then re-queued.

> At this point, the original code invokes kick_object_context_blocked(), then cop->cb->complete() in order.
So, this causes re-queueing blocked op by using enqueue_front(). In other words, kick_object_context_blocked()
enqueue old operation, and cop->cb->complete() enqueue current operation by using enqueue_front(). (note that this is
enqueue_front). But, with this PR, the functions are called in reverse order, which means the stat (as I described above) will be served before handling copy-from because it enqueues the first position in the queue.

> So, I think we can avoid this and resolve the issue this PR posted via #40067 ?

@athanatos @tchaikov ping.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-03-27_12:17:59-rados-wip-kefu-testing-2021-03-27-1545-distro-basic-gibba/
- https://pulpito.ceph.com/kchai-2021-03-28_03:01:25-rados-wip-kefu-testing-2021-03-27-1545-distro-basic-smithi/

failures

- https://tracker.ceph.com/issues/49344
- https://tracker.ceph.com/issues/49260
- https://tracker.ceph.com/issues/49633

**Comment by athanatos** [comment]
Why doesn't the second block in do_op at maybe_await_blocked_head?  I'm worried about this fix, no other promote caller appears to need it and it seems to duplicate the blocked_op machinery that already exists.

**Comment by myoungwon** [comment]
@athanatos Done.

**Comment by athanatos** [comment]
This also doesn't make sense to me.  The problem here is that operations on the head object need to block until the copy op started in rollback_to completes, right?  Can't you set the head obc there to blocked when you kick off the copy operation?

**Comment by myoungwon** [comment]
Right. I need more time to investigate this.

**Comment by athanatos** [comment]
What does maybe_handle_cache_detail do in that situation?  Shouldn't this be symmetric to the existing cache/tiering case?

**Comment by myoungwon** [comment]
maybe_await_blocked_head() is not the right place to resolve this issue.
Please see the following case where the error occurs.

1. A object evicted (foo:2, foo:3, foo:head).
2. Rollback head to snapid 3.
3. Sparse read to snapid 2

At this time, 3 try to promote the object with snap id 3 because it is missing oid.
So, cancel_copy -> kick_object_context_blocked loop re-appears.

**Comment by athanatos** [comment]
Right.  In step 2, I think head should probably be blocked.  What happens if the pool is a cache/tier pool and snapid 3 is not present in the cache tier?

**Comment by myoungwon** [comment]
`What happens if the pool is a cache/tier pool and snapid 3 is not present in the cache tier?`
Promote_object() to copy snapid 3 ?

**Comment by athanatos** [comment]
I mean in the context of the behaviour you are seeing.  I suggest carefully tracing the existing cache/tiering machinery to find out how (if?) that case is handled for cache/tier pools.  I *really* do not want to introduce another pathway unless the existing one is non-existent or for some specific reason inappropriate.

**Comment by myoungwon** [comment]
Updated. I think the updated commit is more practical way to resolve this issue. This blocks head object using start_block() when the object is rollbacked. 
So, sparse read should be blocked by maybe_await_head()
Then, it calls stop_block() and requeue_ops() to kick blocked ops for the head.

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -711,6 +711,9 @@ void PrimaryLogPG::block_write_on_snap_rollback(
   // otherwise, we'd have blocked in do_op
   ceph_assert(oid.is_head());
   ceph_assert(objects_blocked_on_snap_promotion.count(oid) == 0);
+  // see https://tracker.ceph.com/issues/49726
```
Comment (athanatos) [inline src/osd/PrimaryLogPG.cc:714]: I'd like an actual explanation here.

**Comment by myoungwon** [comment]
Right.

File: src/osd/PrimaryLogPG.cc (hunk 1 above)
Comment (myoungwon) [reply src/osd/PrimaryLogPG.cc]: Updated.

**Comment by athanatos** [comment]
Have you run this through the full suite yet?

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-04-11_12:36:03-rados-wip-kefu-testing-2021-04-11-1826-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-04-12_02:29:32-rados-wip-kefu-testing-2021-04-12-0057-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-04-12_08:01:43-rados-wip-kefu-testing-2021-04-12-1256-distro-basic-smithi/

multi-site failure tracked by https://tracker.ceph.com/issues/50218

**Review by athanatos — APPROVED** [review]

**Comment by batrick** [comment]
dev compiler fails: https://tracker.ceph.com/issues/50342

**Pull request merged**